src/middleware/current_user.rs

```diff
@@ -1,104 +1,37 @@
 
-use crate::models::ApiToken;
-use crate::models::User;
-use crate::schema::users;
+/// A trusted user_id extracted from a signed cookie or added to the request by the test harness
+#[derive(Clone, Copy, Debug)]
+pub struct TrustedUserId(pub i32);
 
-#[derive(Debug, Clone, Copy)]
-pub struct CurrentUser;
+/// Middleware that captures the `user_id` from the signed session cookie
+pub(super) struct CaptureUserIdFromCookie;
 
-#[derive(Debug, Copy, Clone, Eq, PartialEq)]
-pub enum AuthenticationSource {
-    SessionCookie,
-    ApiToken { api_token_id: i32 },
-}
-
-impl Middleware for CurrentUser {
+impl Middleware for CaptureUserIdFromCookie {
     fn before(&self, req: &mut dyn Request) -> Result<(), Box<dyn Error + Send>> {
-        // Check if the request has a session cookie with a `user_id` property inside
-        let id = {
-            req.session()
-                .get("user_id")
-                .and_then(|s| s.parse::<i32>().ok())
-        };
-
-        let conn = req.db_conn().map_err(std_error)?;
-
-        if let Some(id) = id {
-            // If it did, look for a user in the database with the given `user_id`
-            let maybe_user = users::table.find(id).first::<User>(&*conn);
-            drop(conn);
-            if let Ok(user) = maybe_user {
-                // Attach the `User` model from the database to the request
-                req.mut_extensions().insert(user);
-                req.mut_extensions()
-                    .insert(AuthenticationSource::SessionCookie);
-            }
-        } else {
-            // Otherwise, look for an `Authorization` header on the request
-            // and try to find a user in the database with a matching API token
-            let user_auth = if let Some(headers) = req.headers().find("Authorization") {
-                ApiToken::find_by_api_token(&conn, headers[0])
-                    .and_then(|api_token| {
-                        User::find(&conn, api_token.user_id).map(|user| {
-                            (
-                                AuthenticationSource::ApiToken {
-                                    api_token_id: api_token.id,
-                                },
-                                user,
-                            )
-                        })
-                    })
-                    .optional()
-                    .map_err(|e| Box::new(e) as Box<dyn Error + Send>)?
-            } else {
-                None
-            };
-
-            drop(conn);
-
-            if let Some((api_token, user)) = user_auth {
-                // Attach the `User` model from the database and the API token to the request
-                req.mut_extensions().insert(user);
-                req.mut_extensions().insert(api_token);
-            }
+        if let Some(id) = req.session().get("user_id").and_then(|s| s.parse().ok()) {
+            req.mut_extensions().insert(TrustedUserId(id));
```

Comment:
out of curiosity, why is it not possible to put this into the authenticate call too?

Reply:
> I think this req.session() is the call that needs the mutable reference to the request

Yes, that's exactly the problem I ran into.  `session()` would probably be better called `session_mut()`, since it calls `mut_extensions()` and `find_mut()` [internally](https://github.com/conduit-rust/conduit-cookie/blob/master/src/session.rs#L100-L105) to return a `&mut HashMap<_, _>`.